WalletConnectSharp.Sign/Controllers/AddressProvider.cs

```diff
@@ -75,45 +140,52 @@ private void ClientOnSessionConnected(object sender, SessionStruct e)
-            if (currentDefault != null && DefaultSession.RequiredNamespaces.ContainsKey(currentDefault))
+            if (HasDefaultSession)
             {
-                // DefaultNamespace is still valid
-                var currentChain = DefaultChain;
-                if (currentChain == null ||
-                    DefaultSession.RequiredNamespaces[DefaultNamespace].Chains.Contains(currentChain))
+                var currentDefault = DefaultNamespace;
+                if (currentDefault != null && DefaultSession.RequiredNamespaces.ContainsKey(currentDefault))
                 {
-                    // DefaultChain is still valid
+                    // DefaultNamespace is still valid
+                    var currentChain = DefaultChain;
+                    if (currentChain == null ||
+                        DefaultSession.RequiredNamespaces[DefaultNamespace].Chains.Contains(currentChain))
+                    {
+                        // DefaultChain is still valid
+                        return;
+                    }
+
+                    DefaultChain = DefaultSession.RequiredNamespaces[DefaultNamespace].Chains[0];
                     return;
                 }
 
-                DefaultChain = DefaultSession.RequiredNamespaces[DefaultNamespace].Chains[0];
-                return;
-            }
-
-            // DefaultNamespace is null or not found in RequiredNamespaces, update it
-            DefaultNamespace = DefaultSession.RequiredNamespaces.OrderedKeys.FirstOrDefault();
-            if (DefaultNamespace != null)
-            {
-                DefaultChain = DefaultSession.RequiredNamespaces[DefaultNamespace].Chains[0];
-            }
-            else
-            {
-                // TODO The Keys property is unordered! Maybe this needs to be updated
-                DefaultNamespace = DefaultSession.RequiredNamespaces.Keys.FirstOrDefault();
+                // DefaultNamespace is null or not found in RequiredNamespaces, update it
+                DefaultNamespace = DefaultSession.RequiredNamespaces.OrderedKeys.FirstOrDefault();
                 if (DefaultNamespace != null)
                 {
                     DefaultChain = DefaultSession.RequiredNamespaces[DefaultNamespace].Chains[0];
                 }
+                else
+                {
+                    // TODO The Keys property is unordered! Maybe this needs to be updated
+                    DefaultNamespace = DefaultSession.RequiredNamespaces.Keys.FirstOrDefault();
+                    if (DefaultNamespace != null)
+                    {
+                        DefaultChain = DefaultSession.RequiredNamespaces[DefaultNamespace].Chains[0];
+                    }
+                }
+            }
+            else
+            {
+                DefaultNamespace = null;
             }
         }
-        else
+        finally
         {
-            DefaultNamespace = null;
+            await SaveDefaults();
```

Comment:
Why is this in `finally`? What if `try` block breaks halfway and we save incorrect session?